docker/docker.github.io
# [WIP] Dockerfile per version with nginx as reverse proxy (take2)

Pushing as a separate branch for now, as I couldn't get my Git remote setup correctly :blush:

This is still WIP, but adds some modifications to @mstanleyjones's PR https://github.com/docker/docker.github.io/pull/913

- Adding a minimal(ish) base image for the docs to publish
- Updated the Dockerfile and build-script for archive builds
- Updated the nginx-configuration to route /v1.xx/ urls to the right container


To test;

- go to the `_data/docsarchive/` directory
- run `./build_all_archives.sh`
- wait for the builds to complete
- go to the `_data/docsarchive/proxy/` directory
- run `docker-compose up --build -d`
- run `docker-compose logs -f` to watch the logs
- open http://localhost:4000/ in your browser
- open http://localhost:4000/v1.12/ in your browser to view the current docs as well (test if navigation works correctly)
- open http://localhost:4000/v1.4/ for a trip down memory lane

In a terminal;

run `docker images docs`; images should be smaller than previously.Note that 143MB of the size is taken up by the base image (which is shared by all images).

```
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
docs                latest              b1cf8892f1ef        About an hour ago   316 MB
docs                v1.12               1544ee4f8a92        About an hour ago   319 MB
docs                v1.11               112c42e65884        About an hour ago   286 MB
docs                v1.10               88b6585ff788        About an hour ago   270 MB
docs                v1.9                bd84e9b071a6        About an hour ago   211 MB
docs                v1.8                cfc0b2e4ea65        About an hour ago   205 MB
docs                v1.7                adde0f58abb6        About an hour ago   202 MB
docs                v1.6                93b6d13fb81c        About an hour ago   172 MB
docs                v1.5                61d4cafff996        About an hour ago   161 MB
docs                v1.4                dee7a0fef8db        About an hour ago   156 MB
docs                base                ef794735ce91        3 days ago          143 MB
```

In a terminal, run

    docker ps --filter label=com.docker.compose.project=proxy --format '{{.Names}}' | xargs docker stats

To view memory and cpu-use of the containers;

```
CONTAINER                    CPU %               MEM USAGE / LIMIT    MEM %               NET I/O             BLOCK I/O           PIDS
proxy_docs-archive-1-6_1     0.00%               2.004 MiB / 25 MiB   8.02%               3.18 kB / 648 B     205 kB / 0 B        2
proxy_docs-archive-1-7_1     0.00%               1.707 MiB / 25 MiB   6.83%               3.27 kB / 648 B     8.19 kB / 0 B       2
proxy_docs-archive-1-5_1     0.00%               1.699 MiB / 25 MiB   6.80%               4 kB / 648 B        12.3 kB / 0 B       2
proxy_docs-archive-1-9_1     0.00%               1.719 MiB / 25 MiB   6.88%               4 kB / 648 B        12.3 kB / 0 B       2
proxy_docs-archive-1-12_1    0.00%               1.914 MiB / 25 MiB   7.66%               2.44 kB / 648 B     0 B / 0 B           2
proxy_docs-archive-1-4_1     0.00%               2.617 MiB / 25 MiB   10.47%              7.26 kB / 648 B     942 kB / 0 B        2
proxy_main_1                 0.00%               1.746 MiB / 25 MiB   6.98%               6.74 kB / 648 B     20.5 kB / 0 B       2
proxy_docs-archive-1-10_1    0.00%               1.73 MiB / 25 MiB    6.92%               6.74 kB / 648 B     16.4 kB / 0 B       2
proxy_docs-archive-1-11_1    0.00%               1.727 MiB / 25 MiB   6.91%               5.74 kB / 648 B     20.5 kB / 0 B       2
proxy_docs-archive-1-8_1     0.00%               2.34 MiB / 25 MiB    9.36%               7.26 kB / 648 B     659 kB / 0 B        2
proxy_docs-current_1         0.00%               1.719 MiB / 25 MiB   6.88%               5.99 kB / 648 B     20.5 kB / 0 B       2
```

Compared to current "master"

```
CONTAINER           CPU %               MEM USAGE / LIMIT       MEM %               NET I/O             BLOCK I/O           PIDS
docs-orig           0.00%               396.2 MiB / 1.952 GiB   19.82%              648 B / 648 B       212 MB / 0 B        4
docs-master         0.00%               657.2 MiB / 1.952 GiB   32.88%              1.3 kB / 648 B      315 MB / 0 B        4
```
